### PR TITLE
cache redesign: api v3

### PR DESCRIPTION
**Pull request by atolopko-czi**
#### Reviewers
**Functional:** 
@Bento007 

**Readability:** 
@seve 

---

## Changes

This PR supercedes https://github.com/chanzuckerberg/single-cell-explorer/pull/175.  Branched from api-v3 changes and made some additional fixes and refactoring. Added comments for changes worth CR'ing.  Most of this code was CR'd on #175, but this branch also included recent rdev changes that were recently merged in `main` and the `epic/164-cache-redesign` branch and I didn't feel compelled to bring the original `dunitz/api-v3` branch up-to-date as well.  Forgive me for opening a new PR.

Tested on [rdev](https://rdev-164-caching-redesign-frontend.rdev.single-cell.czi.technology/)

### CODE REVIEW

File: hosted/create_rdev_config.py [hunk 1]
```diff
@@ -47,23 +47,6 @@
...
-      timelimit_s: 30
-
-    metadata_cache:
-      # The maximum number of datasets that may be opened at one time.  The least recently used dataset
-      # is evicted from the cache first.
-      max_datasets: 10000
-      # A matrix is automatically removed from the cache after timelimit_s number of seconds.
-      # If timelimit_s is set to None, then there is no time limit.
-      timelimit_s: 300
-
```
Comment (atolopko-czi) [inline hosted/create_rdev_config.py:66]: rdev fix

File: server/data_common/dataset_metadata.py [hunk 2]
```diff
@@ -27,47 +27,47 @@ def request_dataset_metadata_from_data_portal(data_portal_api_base: str, explore
...
+
+    s3_uri = path_join(dataroot, dataset_id)
+
+    # TODO: Is the check still necessary?
+    # path_join returns a normalized path.  Therefore it is
+    # sufficient to check that the datapath starts with the
+    # dataroot to determine that the datapath is under the dataroot.
+    if not s3_uri.startswith(dataroot):
+        raise DatasetAccessError(f"Invalid dataset {s3_uri}")
+    return s3_uri
```
Comment (atolopko-czi) [inline server/data_common/dataset_metadata.py:56]: Refactored and simplified `extrapolate_dataset_location_from_config`, renaming to `infer_dataset_s3_uri`. In particular, split the `dataset_explorer_location` param into separate `dataset_root` and `dataset_id` params. This is intended to make it clearer what specifically the method needs, and these are common params throughout the code base. The `dataset_explorer_location` param was not descriptive enough, and callers were having to construct it from the separate params.

File: server/data_common/dataset_metadata.py [hunk 3]
```diff
@@ -85,30 +85,27 @@ def get_dataset_metadata_for_explorer_location(dataset_explorer_location: str, a
...
         "tombstoned": False,
     }
-    dataset_metadata["s3_uri"] = extrapolate_dataset_location_from_config(
-        server_config=server_config, dataset_explorer_location=dataset_explorer_location
-    )
-    if dataset_metadata["s3_uri"] is None:
-        current_app.logger.log(logging.INFO, f"Dataset not found: {dataset_explorer_location}")
-        raise DatasetNotFoundError(f"Dataset location not found for {dataset_explorer_location}")
-
-    return dataset_metadata
```
Comment (atolopko-czi) [inline server/data_common/dataset_metadata.py:103]: Above is refactoring of `get_dataset_metadata_for_explorer_location()`, renaming to `get_dataset_metadata()` and also now taking the separate `dataset_root` and dataset_id` params for clarity.  Improved logging messages.

File: server/data_common/dataset_metadata.py [hunk 4]
```diff
@@ -85,30 +85,27 @@ def get_dataset_metadata_for_explorer_location(dataset_explorer_location: str, a
...
 
 
-def get_dataset_and_collection_metadata(dataset_explorer_location: str, app_config: AppConfig, current_app):
+def get_dataset_and_collection_metadata(dataset_root: str, dataset_id: str, app_config: AppConfig):
```
Comment (atolopko-czi) [inline server/data_common/dataset_metadata.py:104]: same param splitting here

File: server/data_common/matrix_loader.py [hunk 5]
```diff
@@ -13,21 +13,10 @@ class MatrixDataType(Enum):
...
-            dataroot = None
-            for key, dataroot_dict in self.app_config.server_config.multi_dataset__dataroot.items():
-                if dataroot_dict["base_url"] == self.url_dataroot:
-                    dataroot = key
-                    break
-            if dataroot:
-                self.dataset_config = self.app_config.get_dataset_config(dataroot)
-        if self.dataset_config is None:
-            raise DatasetAccessError("Missing dataset config", HTTPStatus.NOT_FOUND)
+        self.dataset_config = self.__resolve_dataset_config()
```
Comment (atolopko-czi) [inline server/data_common/matrix_loader.py:19]: factored out code into method

File: server/tests/unit/common/config/test_server_config.py [hunk 6]
```diff
@@ -47,7 +47,7 @@ def test_init_raises_error_if_default_config_is_invalid(self):
     def test_complete_config_checks_all_attr(self, mock_check_attrs):
         mock_check_attrs.side_effect = BaseConfig.validate_correct_type_of_configuration_attribute()
         self.server_config.complete_config(self.context)
-        self.assertEqual(mock_check_attrs.call_count, 36)
+        self.assertEqual(mock_check_attrs.call_count, 32)
```
Comment (atolopko-czi) [inline server/tests/unit/common/config/test_server_config.py:50]: There are now 4 less config properties related to removal of caching, so this is a valid test update

**Review by Bento007 — APPROVED** [review]
Wow, that was a lot. In the future if this could be broken into small easier to digest snippets. I see no issues

**Comment by atolopko-czi** [comment]
> Wow, that was a lot. In the future if this could be broken into small easier to digest snippets. I see no issues

Yeah, there was a lot of code that needed to be improved, if only to help me see the problems clearer. Thanks for reviewing!

**Pull request merged**